### PR TITLE
[FSPE-5200] Fixed Tooltip.js Show-Hide behavior

### PR DESCRIPTION
**Pull request by janibolkvadze**
### Commit summary

- [x] Fixed Tooltip.JS `onMouseOut` behavior (for cases of complex triggers like buttons with icons and other elements)
- [x] Added generating of a unique ID for each Tooltip by Introducing `uuid4()` Util
- [x] Moved injecting of Tooltip elements to the end of the page before closing the `</body>` tag

@dani-cl-madrid @SergioQCL @mattnickles @jllr 

https://ctl.atlassian.net/browse/FSPE-5200

### CODE REVIEW

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://d0079c755ff082b541fa86d44396beb047c77472-chi-pr.enterprise.ctl.io" target="_blank"> https://d0079c755ff082b541fa86d44396beb047c77472-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1056/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://7b115c5a2d208b47050bb5e2d6a81c407463e30a-chi-pr.enterprise.ctl.io" target="_blank"> https://7b115c5a2d208b47050bb5e2d6a81c407463e30a-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1062/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://1bf2d4889ac21ebf95c5882cbfeda3410733ea12-chi-pr.enterprise.ctl.io" target="_blank"> https://1bf2d4889ac21ebf95c5882cbfeda3410733ea12-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1063/ for more details.

File: src/chi/javascript/components/tooltip.js [hunk 1]
```diff
@@ -56,7 +56,7 @@ class Tooltip extends Component {
     };
 
     this._elem.addEventListener('mouseover',this._mouseOverEventHandler,false);
-    this._elem.addEventListener('mouseout', this._mouseOutEventHandler, false);
+    this._elem.addEventListener('mouseleave', this._mouseLeaveEventHandler, false);
     this._elem.addEventListener('focus',this._focusEventHandler,false);
```
Comment (SergioQCL) [inline src/chi/javascript/components/tooltip.js:60]: Do you remove also the listener for focus and blur events?

File: src/chi/javascript/components/tooltip.js [hunk 2]
```diff
@@ -92,12 +92,11 @@ class Tooltip extends Component {
   _createTooltip () {
     this._tooltipElem = document.createElement('div');
     this._tooltipElem.setAttribute('class', 'chi-tooltip');
-    this._tooltipElem.id = COMPONENT_TYPE +
-      Util.getData(this._elem, COMPONENT_TYPE);
+    this._tooltipElem.id = 'chi-' + COMPONENT_TYPE + '-' + this.componentCounterNo;
```
Comment (SergioQCL) [inline src/chi/javascript/components/tooltip.js:95]: You can use a string template for this kind of things:
```suggestion
    this._tooltipElem.id = `chi-${COMPONENT_TYPE}-${this.componentCounterNo}`;
```

File: src/chi/javascript/core/util.js [hunk 3]
```diff
@@ -533,6 +533,30 @@ export class Util {
     return null;
   }
 
+  static uuid4() {
+    let uuid = '', ii;
```
Comment (SergioQCL) [inline src/chi/javascript/core/util.js:537]: Please, define each variable in its own line, and give a more meaningful name to _ii_ if possible. Thanks in advance

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://98d9707cdcaedf90cba01c0e2070d972c981a201-chi-pr.enterprise.ctl.io" target="_blank"> https://98d9707cdcaedf90cba01c0e2070d972c981a201-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1065/ for more details.

**Comment by allyourbotsarebelongtous** [comment]
Your CHI Pull Request website is available for review at <a href="https://332a5e38edca9d7516bae66189253f5f35bc8b30-chi-pr.enterprise.ctl.io" target="_blank"> https://332a5e38edca9d7516bae66189253f5f35bc8b30-chi-pr.enterprise.ctl.io </a>

**Comment by allyourbotsarebelongtous** [comment]
Build finished SUCCESSFULLY. See http://jenkins.elasticbox.com/job/chi_pull_request/1070/ for more details.

File: src/chi/javascript/components/tooltip.js [hunk 4]
```diff
@@ -92,12 +87,11 @@ class Tooltip extends Component {
...
     this._tooltipElem = document.createElement('div');
     this._tooltipElem.setAttribute('class', 'chi-tooltip');
-    this._tooltipElem.id = COMPONENT_TYPE +
-      Util.getData(this._elem, COMPONENT_TYPE);
+    this._tooltipElem.id = `chi-${COMPONENT_TYPE}-${this.componentCounterNo}`;
     this._tooltipContent = document.createElement('span');
     this._tooltipContent.innerText = this._elem.dataset.tooltip;
     this._tooltipElem.appendChild(this._tooltipContent);
-    this._elem.parentNode.appendChild(this._tooltipElem);
+    document.querySelector('body').appendChild(this._tooltipElem);
```
Comment (dani-cl-madrid) [inline src/chi/javascript/components/tooltip.js:94]: Please, remove this element from the body when disposing.

Comment (janibolkvadze) [reply src/chi/javascript/components/tooltip.js:94]:
```
this._tooltipElem = null;
this._tooltipContent = null;
```
It is being removed when dispose is triggered

Comment (dani-cl-madrid) [reply src/chi/javascript/components/tooltip.js:94]: You are clearing local variables, but it doesn't mean the element is removed from the body. It is about references. You have deleted the local reference to the element, but the element is still there.

**Comment by jllr** [comment]
@dani-cl-madrid is this PR ready to merge?

**Comment by dani-cl-madrid** [comment]
> @dani-cl-madrid is this PR ready to merge?

Yes. Ready to be merged.

**Comment by allyourbotsarebelongtous** [comment]
The following instances are being terminated: https://ops.elasticbox.com/services/instances/i-jisme1, https://ops.elasticbox.com/services/instances/i-rgtymc, https://ops.elasticbox.com/services/instances/i-52w9h3, https://ops.elasticbox.com/services/instances/i-842019, https://ops.elasticbox.com/services/instances/i-a3t0gn

**Pull request merged**